pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,282 @@
+package plugin
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/Masterminds/semver/v3"
+	filepathsecure "github.com/cyphar/filepath-securejoin"
+	v1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	FilesTxtFilename    = "files.txt"
+	PackageJSONFilename = "plugin/package.json"
+
+	// Cache states used by custom resources
+	Cached   = "cached"
+	Disabled = "disabled"
+	Pending  = "pending"
+)
+
+var (
+	FsCache             = FSCache{}
+	errMaxFileSizeError = fmt.Errorf("file size limit of %s bytes reached", settings.MaxUIPluginFileByteSize.Get())
+	FSCacheRootDir      = filepath.Join("management-state", "uiplugin")
+	osRemoveAll         = os.RemoveAll
+	osStat              = os.Stat
+	isDirEmpty          = isDirectoryEmpty
+)
+
+type FSCache struct {
+}
+
+type PackageJSON struct {
+	Version string `json:"version,omitempty"`
+}
+
+// SyncWithControllersCache takes in a UI Plugin object and syncs the filesystem cache with it
+func (c FSCache) SyncWithControllersCache(p *v1.UIPlugin) error {
+	plugin := p.Spec.Plugin
+	if plugin.NoCache {
+		logrus.Debugf("skipped caching plugin [Name: %s Version: %s] cache is disabled [noCache: %v]", plugin.Name, plugin.Version, plugin.NoCache)
+		return nil
+	}
+	if isCached, err := c.isCached(plugin.Name, plugin.Version); err != nil {
+		return err
+	} else if isCached {
+		logrus.Debugf("skipped caching plugin [Name: %s Version: %s] is already cached", plugin.Name, plugin.Version)
+		return nil
+	}
+	version, err := getVersionFromPackageJSON(fmt.Sprintf("%s/%s", plugin.Endpoint, PackageJSONFilename))
+	if err != nil {
+		return err
+	}
+	cachedVersion, err := semver.NewVersion(plugin.Version)
+	if err != nil {
+		return err
+	}
+	if !cachedVersion.Equal(version) {
+		return fmt.Errorf("plugin [%s] version [%s] does not match version in controller's cache [%s]", plugin.Name, version.String(), cachedVersion.String())
+	}
+	files, err := readFilesTxt(fmt.Sprintf("%s/%s", plugin.Endpoint, FilesTxtFilename))
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		if file == "" {
+			continue
+		}
```

Comment:
Please add some validation against `file`. It must not:
- Start with `/` or `\`.
- Start with `.`.
- Contain `..`.

Given that this is being used for an HTTP request, ensure file is URL decoded ahead of the validation.